File: server/models/backtest.py
"""回测 Pydantic schemas"""
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class BacktestRequest(BaseModel):
    """回测请求参数"""
    code: str = Field(..., description="股票代码")
    start_date: str = Field("2024-01-01", description="回测开始日期 YYYY-MM-DD")
    end_date: Optional[str] = Field(None, description="回测结束日期 YYYY-MM-DD，默认当天")
    initial_capital: float = Field(100000, gt=0, description="初始本金")
    exit_strategy: str = Field("fixed", description="出场策略 fixed/trailing/boll_middle/trailing_boll/half_exit/half_exit_low3")
    tp_multiplier: float = Field(2.0, gt=0, description="止盈倍数")
    trailing_atr_k: float = Field(1.0, ge=0, description="跟踪止损ATR系数")
    half_exit_pct: float = Field(50, ge=0, le=100, description="半仓止盈比例%")


class TradeResult(BaseModel):
    """单笔交易结果"""
    entry_date: str
    exit_date: str
    entry_price: float
    exit_price: float
    stop_loss: float
    take_profit: float
    shares: int
    pnl: float
    pnl_r: float
    holding_days: int
    reason: str
    atr: float
    upper_band: float = 0
    breakout_close: float = 0
    breakout_exceed_pct: float = 0
    exit_formula: str = ""


class EquityPoint(BaseModel):
    """净值曲线数据点"""
    date: str
    total: float
    equity: float
    position_value: float
    peak: float
    dd: float
    dd_pct: float


class KlineBar(BaseModel):
    """K 线数据"""
    date: str
    open: Optional[float] = None
    close: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    volume: Optional[float] = None
    turnover: Optional[float] = None


class BacktestStats(BaseModel):
    """回测统计指标"""
    initial_capital: float
    final_capital: float
    total_return: float
    total_return_pct: float
    num_trades: int
    win_trades: int
    loss_trades: int
    win_rate: float
    avg_win: float
    avg_loss: float
    avg_pnl: float
    rr_ratio: float
    max_drawdown: float
    max_drawdown_pct: float
    max_dd_date: Optional[str] = None
    sharpe_ratio: float
    avg_holding_days: float
    daily_return_pct: float = 0


class BacktestResponse(BaseModel):
    """回测响应"""
    code: str
    name: str
    request: BacktestRequest
    stats: BacktestStats
    trades: List[TradeResult]
    equity_curve: List[EquityPoint]
    klines: List[KlineBar]


class SaveBacktestRequest(BaseModel):
    """保存回测结果请求（复用运行时的请求和响应数据）"""
    code: str
    name: str
    request: BacktestRequest
    stats: BacktestStats
    trades: List[TradeResult]
    equity_curve: List[EquityPoint]
    klines: List[KlineBar]


class BacktestHistoryItem(BaseModel):
    """历史回测列表项"""
    id: int
    code: str
    name: str
    start_date: str
    end_date: str
    exit_strategy: str
    total_return_pct: float
    num_trades: int
    win_rate: float
    max_drawdown_pct: float
    created_at: str


class BacktestHistoryListResponse(BaseModel):
    """历史回测列表响应"""
    items: List[BacktestHistoryItem]
    total: int


class BacktestDetailResponse(BaseModel):
    """历史回测详情响应"""
    id: int
    code: str
    name: str
    start_date: str
    end_date: str
    initial_capital: float
    exit_strategy: str
    tp_multiplier: float
    trailing_atr_k: float
    half_exit_pct: float
    stats: BacktestStats
    trades: List[TradeResult]
    equity_curve: List[EquityPoint]
    klines: List[KlineBar]
    created_at: str


class ExitStrategyInfo(BaseModel):
    """出场策略信息"""
    key: str
    name: str
    description: str = ""


class ExitStrategyListResponse(BaseModel):
    """出场策略列表"""
    strategies: List[ExitStrategyInfo]


# ============================================================
# 组合回测（多股票）
# ============================================================

class PortfolioBacktestRequest(BaseModel):
    """组合回测请求"""
    codes: List[str] = Field(..., description="候选股票池代码列表")
    start_date: str = Field("2024-01-01", description="回测开始日期")
    end_date: Optional[str] = Field(None, description="回测结束日期，默认当天")
    initial_capital: float = Field(100000, gt=0, description="初始本金")
    max_positions: int = Field(3, ge=1, le=10, description="最多同时持仓数")
    exit_strategy: str = Field("fixed", description="出场策略")
    tp_multiplier: float = Field(2.0, gt=0, description="止盈倍数")
    trailing_atr_k: float = Field(1.0, ge=0, description="跟踪止损 ATR 系数")
    half_exit_pct: float = Field(50, ge=0, le=100, description="半仓止盈比例（组合模式保留参数）")
    score_config: Optional[Dict[str, Any]] = Field(None, description="评分配置覆盖")


class StockResult(BaseModel):
    """单只股票在组合中的回测结果"""
    code: str
    name: str
    trades: List[TradeResult]
    equity_curve: List[EquityPoint]
    stats: BacktestStats


class PortfolioBacktestResponse(BaseModel):
    """组合回测响应"""
    portfolio_stats: BacktestStats
    overall_equity: List[EquityPoint]
    stock_results: List[StockResult]


class ScoreDimension(BaseModel):
    """评分维度配置"""
    key: str
    name: str
    weight: float = Field(0, ge=0, le=100)
    enabled: bool = True
    params: Dict[str, Any] = {}


class ScoreConfigResponse(BaseModel):
    """评分配置响应"""
    dimensions: List[ScoreDimension]


# ---------- 候选股票池 ----------

class PortfolioPool(BaseModel):
    """候选股票池"""
    id: int
    name: str
    codes: List[str]
    code_names: Dict[str, str] = Field(default_factory=dict, description="code→名称映射")
    created_at: str


class PortfolioPoolListResponse(BaseModel):
    """候选池列表响应"""
    items: List[PortfolioPool]


class PortfolioPoolCreate(BaseModel):
    """创建候选池"""
    name: str = Field(..., min_length=1, max_length=50)
    codes: List[str] = Field(..., min_length=1)


class PortfolioPoolUpdate(BaseModel):
    """更新候选池"""
    name: Optional[str] = None
    codes: Optional[List[str]] = None

File: web/src/api/backtest.ts
/** 回测相关 API */
import client from './client'

// ---------- 类型定义 ----------

export interface ExitStrategyInfo {
  key: string
  name: string
  description: string
}

export interface BacktestRequest {
  code: string
  start_date: string
  end_date?: string
  initial_capital: number
  exit_strategy: string
  tp_multiplier: number
  trailing_atr_k: number
  half_exit_pct: number
}

export interface TradeResult {
  entry_date: string
  exit_date: string
  entry_price: number
  exit_price: number
  stop_loss: number
  take_profit: number
  shares: number
  pnl: number
  pnl_r: number
  holding_days: number
  reason: string
  atr: number
  upper_band: number
  breakout_close: number
  breakout_exceed_pct: number
  exit_formula: string
  group_date: string
  turtle_units?: { entry_date: string; entry_price: number; shares: number }[]
  turtle_unit_size?: number
}

export interface EquityPoint {
  date: string
  total: number
  equity: number
  position_value: number
  peak: number
  dd: number
  dd_pct: number
}

export interface KlineBar {
  date: string
  open: number | null
  close: number | null
  high: number | null
  low: number | null
  volume: number | null
  turnover: number | null
}

export interface BacktestStats {
  initial_capital: number
  final_capital: number
  total_return: number
  total_return_pct: number
  num_trades: number
  win_trades: number
  loss_trades: number
  win_rate: number
  avg_win: number
  avg_loss: number
  avg_pnl: number
  rr_ratio: number
  max_drawdown: number
  max_drawdown_pct: number
  max_dd_date: string | null
  sharpe_ratio: number
  avg_holding_days: number
  daily_return_pct: number
  best_trade: TradeResult | null
  worst_trade: TradeResult | null
}

export interface BacktestResponse {
  code: string
  name: string
  request: BacktestRequest
  stats: BacktestStats
  trades: TradeResult[]
  equity_curve: EquityPoint[]
  klines: KlineBar[]
}

export interface BacktestHistoryItem {
  id: number
  code: string
  name: string
  start_date: string
  end_date: string
  exit_strategy: string
  total_return_pct: number
  num_trades: number
  win_rate: number
  max_drawdown_pct: number
  created_at: string
}

export interface BacktestHistoryListResponse {
  items: BacktestHistoryItem[]
  total: number
}

export interface BacktestDetailResponse {
  id: number
  code: string
  name: string
  start_date: string
  end_date: string
  initial_capital: number
  exit_strategy: string
  tp_multiplier: number
  trailing_atr_k: number
  half_exit_pct: number
  stats: BacktestStats
  trades: TradeResult[]
  equity_curve: EquityPoint[]
  klines: KlineBar[]
  created_at: string
}

// ---------- API 调用 ----------

/** 获取出场策略列表 */
export async function getStrategies(): Promise<{ strategies: ExitStrategyInfo[] }> {
  const { data } = await client.get('/backtest/strategies')
  return data
}

/** 获取近期回测过的股票 */
export async function getRecentStocks(limit = 10): Promise<{ stocks: { code: string; name: string }[] }> {
  const { data } = await client.get('/backtest/recent-stocks', { params: { limit } })
  return data
}

/** 运行回测 */
export async function runBacktest(params: BacktestRequest): Promise<BacktestResponse> {
  const { data } = await client.post('/backtest/run', params)
  return data
}

/** 保存回测结果 */
export async function saveBacktest(result: BacktestResponse): Promise<{ id: number; message: string }> {
  const { data } = await client.post('/backtest/save', result)
  return data
}

/** 获取历史回测列表 */
export async function getBacktestHistory(params?: {
  page?: number
  page_size?: number
}): Promise<BacktestHistoryListResponse> {
  const { data } = await client.get('/backtest/history', { params })
  return data
}

/** 获取历史回测详情 */
export async function getBacktestDetail(id: number): Promise<BacktestDetailResponse> {
  const { data } = await client.get(`/backtest/history/${id}`)
  return data
}

/** 删除历史回测 */
export async function deleteBacktest(id: number): Promise<void> {
  await client.delete(`/backtest/history/${id}`)
}


// ---------- 组合回测（多股票）----------

export interface PortfolioBacktestRequest {
  codes: string[]
  start_date: string
  end_date?: string
  initial_capital: number
  max_positions: number
  exit_strategy: string
  tp_multiplier: number
  trailing_atr_k: number
  half_exit_pct: number
  score_config?: Record<string, { weight?: number; enabled?: boolean; params?: Record<string, any> }>
}

export interface StockResult {
  code: string
  name: string
  trades: TradeResult[]
  equity_curve: EquityPoint[]
  stats: BacktestStats
}

export interface PortfolioEquityPoint extends EquityPoint {
  num_positions?: number
}

export interface PortfolioBacktestResponse {
  portfolio_stats: BacktestStats
  overall_equity: PortfolioEquityPoint[]
  stock_results: StockResult[]
}

export interface ScoreDimension {
  key: string
  name: string
  weight: number
  enabled: boolean
  params: Record<string, any>
}

export interface ScoreConfigResponse {
  dimensions: ScoreDimension[]
}

/** 运行组合回测 */
export async function runPortfolioBacktest(
  params: PortfolioBacktestRequest
): Promise<PortfolioBacktestResponse> {
  const { data } = await client.post('/backtest/portfolio/run', params)
  return data
}

/** 获取评分配置 */
export async function getScoreConfig(): Promise<ScoreConfigResponse> {
  const { data } = await client.get('/backtest/score-config')
  return data
}

/** 更新评分配置 */
export async function updateScoreConfig(
  dimensions: ScoreDimension[]
): Promise<ScoreConfigResponse> {
  const { data } = await client.post('/backtest/score-config', dimensions)
  return data
}

// ---------- 候选股票池 ----------

export interface PortfolioPool {
  id: number
  name: string
  codes: string[]
  code_names: Record<string, string>
  created_at: string
}

export interface PortfolioPoolListResponse {
  items: PortfolioPool[]
}

/** 获取候选池列表 */
export async function getPortfolioPools(): Promise<PortfolioPoolListResponse> {
  const { data } = await client.get('/backtest/pools')
  return data
}

/** 创建候选池 */
export async function createPortfolioPool(params: { name: string; codes: string[] }): Promise<PortfolioPool> {
  const { data } = await client.post('/backtest/pools', params)
  return data
}

/** 更新候选池 */
export async function updatePortfolioPool(id: number, params: { name?: string; codes?: string[] }): Promise<PortfolioPool> {
  const { data } = await client.put(`/backtest/pools/${id}`, params)
  return data
}

/** 删除候选池 */
export async function deletePortfolioPool(id: number): Promise<void> {
  await client.delete(`/backtest/pools/${id}`)
}
